### ISSUE TITLE
`Dataset.__getitem__` documenting return type

### ISSUE THREAD

**Issue by jamesbraza** [pytorch/pytorch#102907]
### 📚 The doc issue

From https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset, it's unclear what is more optimal: to return `numpy.ndarray` or to return `torch.Tensor`.

I can see `torch.utils.data.default_collate` will convert from `numpy.ndarray` to `torch.Tensor`.

The underlying question I am looking to answer is, for the most memory-efficient pipeline, should I:
- Convert from `numpy.ndarray` to `torch.Tensor` within `Dataset.__getitem__`
- Return `numpy.ndarray` from `Dataset.__getitem__`, and let PyTorch handle the rest

### Suggest a potential alternative/fix

A small section talking about what type to return from `Dataset.__getitem__`.  And if there's tradeoffs, maybe describe that too.

**Comment by NivekT** [comment]
> @NivekT should this go in the torchdata repo instead?

No, this documentation should be in core, because `Dataset` is provided by core.

In terms of documentation, users can easily pass in a custom `collate_fn` to `DataLoader`, such that `default_collate` isn't always used. 

In terms of performance, it should matter too much, because you are essentially perform the same operations in slightly different order. If you benchmark it and find surprising results, feel free to let us know and we can discuss.

**Comment by NivekT** [comment]
> I think there should be a suggested return type 

We intentionally omit that because `Dataset` was not design to make users return any particular data type. Its specific interaction with `collate_fn` is left up to the user to define (if `default_collate` isn't sufficient).

**Comment by jamesbraza** [comment]
Okay I follow, and I think that statement answers my question!  The `__getitem__` should be designed to flow into the collate function, and the collate function's output is what matters.

If I opened a PR for something like this, would you be open to it?
